backend/handler_middleware.go

```diff
@@ -0,0 +1,177 @@
+package backend
+
+import (
+	"context"
+	"errors"
+)
+
+var (
+	errNilRequest = errors.New("req cannot be nil")
+	errNilSender  = errors.New("sender cannot be nil")
+)
+
+// HandlerMiddleware is an interface representing the ability to create a middleware
+// that implements the Handler interface.
+type HandlerMiddleware interface {
+	// CreateHandlerMiddleware creates a new Handler by decorating next Handler.
+	CreateHandlerMiddleware(next Handler) Handler
+}
+
+// The HandlerMiddlewareFunc type is an adapter to allow the use of ordinary
+// functions as HandlerMiddleware's. If f is a function with the appropriate
+// signature, HandlerMiddlewareFunc(f) is a HandlerMiddleware that calls f.
+type HandlerMiddlewareFunc func(next Handler) Handler
+
+// CreateHandlerMiddleware implements the HandlerMiddleware interface.
+func (fn HandlerMiddlewareFunc) CreateHandlerMiddleware(next Handler) Handler {
+	return fn(next)
+}
+
+// MiddlewareHandler decorates a Handler with HandlerMiddleware's.
+type MiddlewareHandler struct {
+	handler Handler
+}
+
+// HandlerFromMiddlewares creates a new MiddlewareHandler implementing Handler that decorates finalHandler with middlewares.
+func HandlerFromMiddlewares(finalHandler Handler, middlewares ...HandlerMiddleware) (*MiddlewareHandler, error) {
+	if finalHandler == nil {
+		return nil, errors.New("finalHandler cannot be nil")
+	}
+
+	return &MiddlewareHandler{
+		handler: handlerFromMiddlewares(middlewares, finalHandler),
+	}, nil
+}
+
+func (h *MiddlewareHandler) setupContext(ctx context.Context, pluginCtx PluginContext, endpoint Endpoint) context.Context {
+	ctx = initErrorSource(ctx)
+	ctx = WithEndpoint(ctx, endpoint)
+	ctx = WithPluginContext(ctx, pluginCtx)
+	ctx = WithGrafanaConfig(ctx, pluginCtx.GrafanaConfig)
+	ctx = WithUser(ctx, pluginCtx.User)
+	ctx = WithUserAgent(ctx, pluginCtx.UserAgent)
```

Comment:
Decided to throw this in here in lack of a better place to ensure context is setup properly for plugins, grafana or anyone that uses middlewares. This was earlier party in the https://github.com/grafana/grafana/blob/6a3dbe7d419ea7b3a9048169c819a16fd8082b47/pkg/plugins/manager/client/decorator.go#L38-L40

Reply:
is your plan to remove setupHandlerContext from [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/adapter_utils.go#L37)?

Reply:
Yes. The plan is to remove/change the adapter_utils to use middlewares instead. Do you see any problem with this?

Reply:
nope, sound fine, just checking what's the plan.